
syntax = "proto2";

package org.nasa.jpl.nexus.ingest.wiretypes;
message GridTile{

    optional ShapedArray latitude = 1;
    optional ShapedArray longitude = 2;

    optional int64 time = 3;

    optional ShapedArray variable_data = 4;

    repeated MetaData meta_data = 5;

}

message SwathTile{

    optional ShapedArray latitude = 1;
    optional ShapedArray longitude = 2;
    optional ShapedArray time = 3;

    optional ShapedArray variable_data = 4;

    repeated MetaData meta_data = 5;

}

message TimeSeriesTile{

    optional ShapedArray latitude = 1;
    optional ShapedArray longitude = 2;

    optional int64 time = 3;

    optional ShapedArray variable_data = 4;

    repeated MetaData meta_data = 5;

}

message MetaData{
    required string name = 1;
    required ShapedArray meta_data = 2;
}

message ShapedArray{
    repeated int32 shape = 1 [packed=true];
    required string dtype = 2;
    required bytes array_data = 3;
}

message Attribute{
    optional string name = 1;
    repeated string values = 2;
}

message TileSummary{

    optional string tile_id = 1;
    optional string section_spec = 2;
    optional string dataset_name = 3;
    optional string granule = 4;
    optional string dataset_uuid = 5;
    optional string data_var_name = 6;
    repeated Attribute global_attributes = 7;

    message BBox{
        optional float lat_min = 1;
        optional float lat_max = 2;
        optional float lon_min = 3;
        optional float lon_max = 4;
    }
    optional BBox bbox = 8;
    
    message DataStats{
        optional float min = 1;
        optional float max = 2;
        optional float mean = 3;
        optional int64 count = 4;

        optional int64 min_time = 5;
        optional int64 max_time = 6;
    }
    optional DataStats stats = 9;
}

message NexusTile{
    optional TileSummary summary = 1;
    optional TileData tile = 2;
}

message TileData{
    optional string tile_id = 1;

    oneof tile_type{
        GridTile grid_tile = 2;
        SwathTile swath_tile = 3;
        TimeSeriesTile time_series_tile = 4;
    }
}

message DimensionSlice{

    oneof name_or_position{
        string dimension_name = 1;
        int32 dimension_position = 2;
    }
    required int64 start = 3;
    required int64 stop = 4;
    optional int64 step = 5;
}